File: src/parsing/expressions/collect_subquery.ts
import SubqueryExpression from "./subquery_expression";

/** `COLLECT { ... }` — the subquery's single returned column gathered into an array. */
class CollectSubquery extends SubqueryExpression {
    protected reduce(rows: any[]): any[] {
        if (rows.length === 0) return [];
        const keys = Object.keys(rows[0]);
        if (keys.length !== 1) {
            throw new Error("COLLECT subquery must return exactly one column");
        }
        const key = keys[0];
        return rows.map((r: any) => r[key]);
    }
}

export default CollectSubquery;

File: flowquery-py/src/parsing/expressions/collect_subquery.py
"""COLLECT { ... } subquery expression."""

from typing import Any, List

from .subquery_expression import SubqueryExpression


class CollectSubquery(SubqueryExpression):
    """`COLLECT { ... }` -- the single returned column gathered into a list."""

    def reduce(self, rows: List[Any], count: int) -> Any:
        if not rows:
            return []
        keys = list(rows[0].keys())
        if len(keys) != 1:
            raise ValueError("COLLECT subquery must return exactly one column")
        key = keys[0]
        return [r[key] for r in rows]
